src/lib.rs

```diff
@@ -1,287 +1,284 @@
-    /// Deploys subnet actor with the corresponding parameters.
-    fn constructor(params: ConstructParams) -> anyhow::Result<Option<RawBytes>>;
-    /// Logic for new peers to join a subnet.
-    fn join(params: JoinParams) -> anyhow::Result<Option<RawBytes>>;
-    /// Called by peers to leave a subnet.
-    fn leave() -> anyhow::Result<Option<RawBytes>>;
-    /// Sends a kill signal for the subnet to the SCA.
-    fn kill() -> anyhow::Result<Option<RawBytes>>;
-    /// Submits a new checkpoint for the subnet.
-    fn submit_checkpoint(ch: Checkpoint) -> anyhow::Result<Option<RawBytes>>;
-}
-
 pub struct Actor;
 
-impl SubnetActor for Actor {
+impl Actor {
     /// The constructor populates the initial state.
     ///
     /// Method num 1. This is part of the Filecoin calling convention.
     /// InitActor#Exec will call the constructor on method_num = 1.
-    fn constructor(params: ConstructParams) -> anyhow::Result<Option<RawBytes>> {
-        // This constant should be part of the SDK.
-        const INIT_ACTOR_ADDR: ActorID = 1;
+    fn constructor<BS, RT>(rt: &mut RT, params: ConstructParams) -> Result<(), ActorError>
+    where
+        BS: Blockstore,
+        RT: Runtime<BS>,
+    {
+        rt.validate_immediate_caller_is(std::iter::once(&*INIT_ACTOR_ADDR))?;
 
-        // Should add SDK sugar to perform ACL checks more succinctly.
-        // i.e. the equivalent of the validate_* builtin-actors runtime methods.
-        // https://github.com/filecoin-project/builtin-actors/blob/master/actors/runtime/src/runtime/fvm.rs#L110-L146
-        let is_test = State::is_test();
-        if sdk::message::caller() != INIT_ACTOR_ADDR
-            && (sdk::message::caller() != TEST_ADDR_ID && is_test)
-        {
-            abort!(USR_FORBIDDEN, "constructor invoked by non-init actor");
-        }
+        let st = State::new(rt.store(), params).map_err(|e| {
+            e.downcast_default(ExitCode::USR_ILLEGAL_STATE, "Failed to create actor state")
+        })?;
 
-        let state = State::new(params, is_test);
-        state.save();
-        Ok(None)
+        rt.create(&st)?;
+
+        Ok(())
     }
 
     /// Called by peers looking to join a subnet.
     ///
     /// It implements the basic logic to onboard new peers to the subnet.
-    fn join(params: JoinParams) -> anyhow::Result<Option<RawBytes>> {
-        let mut st = State::load();
-        let caller = Address::new_id(sdk::message::caller());
-        // check type of caller
-        let code_cid = get_actor_code_cid(&caller).unwrap_or(Cid::default());
-        if sdk::actor::get_builtin_actor_type(&code_cid) != Some(Type::Account) {
-            abort!(USR_FORBIDDEN, "caller not account actor type");
-        }
+    fn join<BS, RT>(rt: &mut RT, params: JoinParams) -> Result<Option<RawBytes>, ActorError>
+    where
+        BS: Blockstore,
+        RT: Runtime<BS>,
+    {
+        let caller = rt.message().caller();
+        // TODO: shall we check caller interface instead here?
```

Comment:
Let's add TODO that we should check here that this is an EOA. I guess we only want to support EOA here for now.